File: backend/verifier_inputs.py
"""Canonical verifier-input codec for Harpocrates (codec ``hpx-vi/1``).

This module is the Python side of a three-way codec that must agree byte for
byte with:

    frontend/src/verifierInputs.ts                       (browser / TypeScript)
    contracts/contracts/harpocrates-registry/src/lib.rs  (Soroban / Rust)

Agreement is enforced by the shared corpus in
``zk/vectors/verifier_conformance_v1.json``; see docs/zk-conformance-vectors.md.

Design rules
------------
* **Bounded.** Every entry point rejects oversized material *before* allocating
  a decoded buffer, so hostile inputs cannot drive memory use.
* **Deterministic.** Validation runs in a fixed order and returns the first
  failing :class:`RejectCode`. The same bytes always produce the same code on
  every layer.
* **Silent.** Errors carry a stable machine code and a field *name* only. No
  witness material, proof bytes, or public-input bytes ever reach the message,
  so the error is safe to log and safe to return to a caller.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Final

CODEC_ID: Final[str] = "hpx-vi/1"

FIELD_LEN: Final[int] = 32
FIELD_COUNT: Final[int] = 4
PUBLIC_INPUTS_LEN: Final[int] = FIELD_LEN * FIELD_COUNT

MIN_PROOF_BYTES: Final[int] = 64
MAX_PROOF_BYTES: Final[int] = 65_536

#: Hard ceiling on accepted hex input length. Anything longer is rejected on the
#: string itself, before any decoding, so a hostile caller cannot force a large
#: allocation just to be told the value is too big.
MAX_HEX_CHARS: Final[int] = 2 * (MAX_PROOF_BYTES + 1_024)

#: BN254 scalar field modulus. A 32-byte big-endian encoding is canonical only
#: when the value it denotes is strictly below this.
BN254_SCALAR_FIELD_MODULUS: Final[int] = (
    21888242871839275222246405745257275088548364400416034343698204186575808495617
)

#: Byte-for-byte identical to ``REVOCATION_DOMAIN_SEPARATOR`` in the Soroban
#: registry: seven bytes of BN254 padding followed by 25 ASCII bytes.
REVOCATION_DOMAIN_SEPARATOR: Final[bytes] = (b"\x00" * 7) + b"HARPOCRATES_REVOCATION_V1"

SCHEMA_SILENT_WITNESS: Final[str] = "silent_witness/v1"
SCHEMA_REVOCATION_WITNESS: Final[str] = "revocation_witness/v1"

_HEX_DIGITS: Final[frozenset[str]] = frozenset("0123456789abcdefABCDEF")


class RejectCode(str, Enum):
    """Stable rejection codes shared across circuit, backend, browser, chain."""

    MALFORMED_HEX = "malformed_hex"
    LENGTH = "length"
    PADDING = "padding"
    NON_CANONICAL_FIELD = "non_canonical_field"
    ZERO_FIELD = "zero_field"
    DOMAIN_MISMATCH = "domain_mismatch"
    PROOF_UNDERSIZE = "proof_undersize"
    PROOF_OVERSIZE = "proof_oversize"
    UNKNOWN_SCHEMA = "unknown_schema"


class VerifierInputError(ValueError):
    """Raised for any rejected verifier input.

    Carries the machine-readable :class:`RejectCode` and, where meaningful, the
    *name* of the offending field. Never carries field contents.
    """

    def __init__(self, code: RejectCode, field: str | None = None) -> None:
        self.code = code
        self.field = field
        super().__init__(f"{code.value}" if field is None else f"{code.value}:{field}")

    def signal(self) -> dict[str, str]:
        """Privacy-safe structured payload for logs and metrics."""
        payload = {"codec": CODEC_ID, "reject_code": self.code.value}
        if self.field is not None:
            payload["field"] = self.field
        return payload


@dataclass(frozen=True)
class SilentWitnessInputs:
    """Parsed ``silent_witness/v1`` public inputs."""

    video_hash: bytes
    credential_root: bytes
    nullifier: bytes


@dataclass(frozen=True)
class RevocationWitnessInputs:
    """Parsed ``revocation_witness/v1`` public inputs."""

    revocation_root: bytes
    nullifier: bytes
    domain_separator: bytes
    credential_root: bytes


# ── Primitives ──────────────────────────────────────────────────────────────


def decode_hex(
    value: str,
    *,
    field: str = "value",
    oversize_code: RejectCode = RejectCode.PROOF_OVERSIZE,
) -> bytes:
    """Decode an even-length, bounded hex string.

    Rejects odd lengths, non-hex characters, and anything past
    :data:`MAX_HEX_CHARS` without decoding it.
    """
    if not isinstance(value, str):
        raise VerifierInputError(RejectCode.MALFORMED_HEX, field)
    if len(value) > MAX_HEX_CHARS:
        raise VerifierInputError(oversize_code, field)
    if len(value) % 2 != 0:
        raise VerifierInputError(RejectCode.MALFORMED_HEX, field)
    if any(character not in _HEX_DIGITS for character in value):
        raise VerifierInputError(RejectCode.MALFORMED_HEX, field)
    try:
        return bytes.fromhex(value)
    except ValueError as exc:  # pragma: no cover - guarded above
        raise VerifierInputError(RejectCode.MALFORMED_HEX, field) from exc


def is_canonical_field(element: bytes) -> bool:
    """Is this 32-byte big-endian encoding strictly below the BN254 modulus?"""
    return (
        len(element) == FIELD_LEN
        and int.from_bytes(element, "big") < BN254_SCALAR_FIELD_MODULUS
    )


def check_proof_bounds(proof: bytes) -> None:
    """Enforce the accepted proof-blob size window."""
    if len(proof) < MIN_PROOF_BYTES:
        raise VerifierInputError(RejectCode.PROOF_UNDERSIZE, "proof")
    if len(proof) > MAX_PROOF_BYTES:
        raise VerifierInputError(RejectCode.PROOF_OVERSIZE, "proof")


def _split_fields(public_inputs: bytes) -> list[bytes]:
    if len(public_inputs) != PUBLIC_INPUTS_LEN:
        raise VerifierInputError(RejectCode.LENGTH, "public_inputs")
    return [
        public_inputs[index * FIELD_LEN : (index + 1) * FIELD_LEN]
        for index in range(FIELD_COUNT)
    ]


def _require_canonical(fields: list[bytes], names: tuple[str, ...]) -> None:
    for element, name in zip(fields, names):
        if not is_canonical_field(element):
            raise VerifierInputError(RejectCode.NON_CANONICAL_FIELD, name)


def _require_non_zero(field_value: bytes, name: str) -> None:
    if field_value == b"\x00" * FIELD_LEN:
        raise VerifierInputError(RejectCode.ZERO_FIELD, name)


def _require_half_padding(field_value: bytes, name: str) -> bytes:
    """A 128-bit half is carried in the low 16 bytes; the high 16 must be zero."""
    if field_value[:16] != b"\x00" * 16:
        raise VerifierInputError(RejectCode.PADDING, name)
    return field_value[16:]


# ── Schema parsers ──────────────────────────────────────────────────────────

_SILENT_WITNESS_FIELDS: Final[tuple[str, ...]] = (
    "video_hash_hi",
    "video_hash_lo",
    "credential_root",
    "nullifier",
)

_REVOCATION_FIELDS: Final[tuple[str, ...]] = (
    "revocation_root",
    "nullifier",
    "domain_separator",
    "credential_root",
)


def parse_silent_witness_inputs(public_inputs: bytes) -> SilentWitnessInputs:
    """Parse ``silent_witness/v1`` public inputs in canonical check order."""
    fields = _split_fields(public_inputs)

    high = _require_half_padding(fields[0], "video_hash_hi")
    low = _require_half_padding(fields[1], "video_hash_lo")

    _require_canonical(fields, _SILENT_WITNESS_FIELDS)

    _require_non_zero(fields[2], "credential_root")
    _require_non_zero(fields[3], "nullifier")

    return SilentWitnessInputs(
        video_hash=high + low,
        credential_root=fields[2],
        nullifier=fields[3],
    )


def parse_revocation_witness_inputs(public_inputs: bytes) -> RevocationWitnessInputs:
    """Parse ``revocation_witness/v1`` public inputs in canonical check order."""
    fields = _split_fields(public_inputs)

    _require_canonical(fields, _REVOCATION_FIELDS)

    _require_non_zero(fields[0], "revocation_root")
    _require_non_zero(fields[1], "nullifier")
    _require_non_zero(fields[3], "credential_root")

    if fields[2] != REVOCATION_DOMAIN_SEPARATOR:
        raise VerifierInputError(RejectCode.DOMAIN_MISMATCH, "domain_separator")

    return RevocationWitnessInputs(
        revocation_root=fields[0],
        nullifier=fields[1],
        domain_separator=fields[2],
        credential_root=fields[3],
    )


def parse_public_inputs(
    schema: str, public_inputs: bytes
) -> SilentWitnessInputs | RevocationWitnessInputs:
    """Dispatch to the parser for ``schema``."""
    if schema == SCHEMA_SILENT_WITNESS:
        return parse_silent_witness_inputs(public_inputs)
    if schema == SCHEMA_REVOCATION_WITNESS:
        return parse_revocation_witness_inputs(public_inputs)
    raise VerifierInputError(RejectCode.UNKNOWN_SCHEMA, "schema")


# ── Conformance entry point ─────────────────────────────────────────────────


def classify(schema: str, public_inputs_hex: str, proof_hex: str) -> str | None:
    """Classify one conformance case.

    Returns ``None`` when the material is accepted, otherwise the stable
    reject-code string. The check order — public inputs first, then the proof
    blob — is part of the codec contract and is mirrored by every layer.
    """
    try:
        public_inputs = decode_hex(
            public_inputs_hex,
            field="public_inputs",
            oversize_code=RejectCode.LENGTH,
        )
        parse_public_inputs(schema, public_inputs)
        proof = decode_hex(proof_hex, field="proof")
        check_proof_bounds(proof)
    except VerifierInputError as error:
        return error.code.value
    return None

File: frontend/src/verifierInputs.ts
/**
 * Canonical verifier-input codec for Harpocrates (codec `hpx-vi/1`).
 *
 * Browser/TypeScript side of a three-way codec that must agree byte for byte
 * with:
 *
 *   backend/verifier_inputs.py                          (Python)
 *   contracts/contracts/harpocrates-registry/src/lib.rs (Soroban / Rust)
 *
 * Agreement is enforced by the shared corpus in
 * `zk/vectors/verifier_conformance_v1.json`; see docs/zk-conformance-vectors.md.
 *
 * Every entry point is bounded, deterministic, and silent: rejections carry a
 * stable machine code and at most a field *name* — never witness material,
 * proof bytes, or public-input bytes.
 */

export const CODEC_ID = 'hpx-vi/1'

export const FIELD_LEN = 32
export const SILENT_WITNESS_FIELD_COUNT = 4
export const REVOCATION_FIELD_COUNT = 4
export const REDACTION_FIELD_COUNT = 5

export const SILENT_WITNESS_PUBLIC_INPUTS_LEN = FIELD_LEN * SILENT_WITNESS_FIELD_COUNT // 128
export const REVOCATION_PUBLIC_INPUTS_LEN = FIELD_LEN * REVOCATION_FIELD_COUNT // 128
export const REDACTION_PUBLIC_INPUTS_LEN = FIELD_LEN * REDACTION_FIELD_COUNT // 160
export const PUBLIC_INPUTS_LEN = SILENT_WITNESS_PUBLIC_INPUTS_LEN // 128 — used by fuzz harness (smallest schema frame)

export const MIN_PROOF_BYTES = 64
export const MAX_PROOF_BYTES = 65536

/**
 * Hard ceiling on accepted hex input length. Anything longer is rejected on the
 * string itself, before decoding, so a hostile caller cannot force a large
 * allocation just to be told the value is too big.
 */
export const MAX_HEX_CHARS = 2 * (MAX_PROOF_BYTES + 1024)

/** BN254 scalar field modulus. Canonical encodings are strictly below this. */
export const BN254_SCALAR_FIELD_MODULUS =
  21888242871839275222246405745257275088548364400416034343698204186575808495617n

/**
 * Byte-for-byte identical to `REVOCATION_DOMAIN_SEPARATOR` in the Soroban
 * registry: seven bytes of BN254 padding followed by 25 ASCII bytes.
 */
export const REVOCATION_DOMAIN_SEPARATOR_HEX =
  '00000000000000484152504f4352415445535f5245564f434154494f4e5f5631'

/**
 * Byte-for-byte identical to Noir domain tag in redaction lineage circuit:
 * eight bytes of BN254 padding followed by 24 ASCII bytes.
 */
export const REDACTION_WITNESS_DOMAIN_TAG_HEX =
  '0000000000000000484152504f4352415445535f524544414354494f4e5f5631'

export const SCHEMA_SILENT_WITNESS = 'silent_witness/v1'
export const SCHEMA_REVOCATION_WITNESS = 'revocation_witness/v1'
export const SCHEMA_REDACTION_WITNESS = 'redaction_witness/v1'

export type VerifierSchema =
  | typeof SCHEMA_SILENT_WITNESS
  | typeof SCHEMA_REVOCATION_WITNESS
  | typeof SCHEMA_REDACTION_WITNESS

export type RejectCode =
  | 'malformed_hex'
  | 'length'
  | 'padding'
  | 'non_canonical_field'
  | 'zero_field'
  | 'domain_mismatch'
  | 'proof_undersize'
  | 'proof_oversize'
  | 'unknown_schema'

/** Rejection carrying a stable machine code and, at most, a field name. */
export class VerifierInputError extends Error {
  readonly code: RejectCode
  readonly field?: string

  constructor(code: RejectCode, field?: string) {
    super(field === undefined ? code : `${code}:${field}`)
    this.name = 'VerifierInputError'
    this.code = code
    this.field = field
  }

  /** Privacy-safe structured payload for logs and metrics. */
  signal(): Record<string, string> {
    const payload: Record<string, string> = { codec: CODEC_ID, rejectCode: this.code }
    if (this.field !== undefined) {
      payload.field = this.field
    }
    return payload
  }
}

export type SilentWitnessInputs = {
  videoHash: Uint8Array
  credentialRoot: Uint8Array
  nullifier: Uint8Array
}

export type RevocationWitnessInputs = {
  revocationRoot: Uint8Array
  nullifier: Uint8Array
  domainSeparator: Uint8Array
  credentialRoot: Uint8Array
}

export type RedactionWitnessInputs = {
  parentCommitment: Uint8Array
  outputCommitment: Uint8Array
  operationType: Uint8Array
  replayBinding: Uint8Array
  domainTag: Uint8Array
}

const HEX_PATTERN = /^[0-9a-fA-F]*$/

/**
 * Decode an even-length, bounded hex string. Rejects odd lengths, non-hex
 * characters, and anything past {@link MAX_HEX_CHARS} without decoding it.
 */
export function decodeHex(
  value: string,
  field = 'value',
  oversizeCode: RejectCode = 'proof_oversize',
): Uint8Array {
  if (typeof value !== 'string') {
    throw new VerifierInputError('malformed_hex', field)
  }
  if (value.length > MAX_HEX_CHARS) {
    throw new VerifierInputError(oversizeCode, field)
  }
  if (value.length % 2 !== 0 || !HEX_PATTERN.test(value)) {
    throw new VerifierInputError('malformed_hex', field)
  }

  const bytes = new Uint8Array(value.length / 2)
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(value.slice(index * 2, index * 2 + 2), 16)
  }
  return bytes
}

function toBigInt(element: Uint8Array): bigint {
  let accumulator = 0n
  for (const byte of element) {
    accumulator = (accumulator << 8n) | BigInt(byte)
  }
  return accumulator
}

/** Is this 32-byte big-endian encoding strictly below the BN254 modulus? */
export function isCanonicalField(element: Uint8Array): boolean {
  return element.length === FIELD_LEN && toBigInt(element) < BN254_SCALAR_FIELD_MODULUS
}

/** Enforce the accepted proof-blob size window. */
export function checkProofBounds(proof: Uint8Array): void {
  if (proof.length < MIN_PROOF_BYTES) {
    throw new VerifierInputError('proof_undersize', 'proof')
  }
  if (proof.length > MAX_PROOF_BYTES) {
    throw new VerifierInputError('proof_oversize', 'proof')
  }
}

function splitFields(publicInputs: Uint8Array, expectedLen: number, count: number): Uint8Array[] {
  if (publicInputs.length !== expectedLen) {
    throw new VerifierInputError('length', 'public_inputs')
  }
  const fields: Uint8Array[] = []
  for (let index = 0; index < count; index += 1) {
    fields.push(publicInputs.slice(index * FIELD_LEN, (index + 1) * FIELD_LEN))
  }
  return fields
}

function requireCanonical(fields: Uint8Array[], names: readonly string[]): void {
  for (let index = 0; index < fields.length; index += 1) {
    if (!isCanonicalField(fields[index])) {
      throw new VerifierInputError('non_canonical_field', names[index])
    }
  }
}

function requireNonZero(field: Uint8Array, name: string): void {
  if (field.every((byte) => byte === 0)) {
    throw new VerifierInputError('zero_field', name)
  }
}

/** A 128-bit half lives in the low 16 bytes; the high 16 must be zero. */
function requireHalfPadding(field: Uint8Array, name: string): Uint8Array {
  for (let index = 0; index < 16; index += 1) {
    if (field[index] !== 0) {
      throw new VerifierInputError('padding', name)
    }
  }
  return field.slice(16)
}

function concat(high: Uint8Array, low: Uint8Array): Uint8Array {
  const out = new Uint8Array(high.length + low.length)
  out.set(high, 0)
  out.set(low, high.length)
  return out
}

const SILENT_WITNESS_FIELDS = [
  'video_hash_hi',
  'video_hash_lo',
  'credential_root',
  'nullifier',
] as const

const REVOCATION_FIELDS = [
  'revocation_root',
  'nullifier',
  'domain_separator',
  'credential_root',
] as const

const REDACTION_FIELDS = [
  'parent_commitment',
  'output_commitment',
  'operation_type',
  'replay_binding',
  'domain_tag',
] as const

/** Parse `silent_witness/v1` public inputs in canonical check order. */
export function parseSilentWitnessInputs(publicInputs: Uint8Array): SilentWitnessInputs {
  const fields = splitFields(publicInputs, SILENT_WITNESS_PUBLIC_INPUTS_LEN, SILENT_WITNESS_FIELD_COUNT)

  const high = requireHalfPadding(fields[0], 'video_hash_hi')
  const low = requireHalfPadding(fields[1], 'video_hash_lo')

  requireCanonical(fields, SILENT_WITNESS_FIELDS)

  requireNonZero(fields[2], 'credential_root')
  requireNonZero(fields[3], 'nullifier')

  return {
    videoHash: concat(high, low),
    credentialRoot: fields[2],
    nullifier: fields[3],
  }
}

/** Parse `revocation_witness/v1` public inputs in canonical check order. */
export function parseRevocationWitnessInputs(
  publicInputs: Uint8Array,
): RevocationWitnessInputs {
  const fields = splitFields(publicInputs, REVOCATION_PUBLIC_INPUTS_LEN, REVOCATION_FIELD_COUNT)

  requireCanonical(fields, REVOCATION_FIELDS)

  requireNonZero(fields[0], 'revocation_root')
  requireNonZero(fields[1], 'nullifier')
  requireNonZero(fields[3], 'credential_root')

  const expectedDomain = decodeHex(REVOCATION_DOMAIN_SEPARATOR_HEX, 'domain_separator')
  const domain = fields[2]
  for (let index = 0; index < FIELD_LEN; index += 1) {
    if (domain[index] !== expectedDomain[index]) {
      throw new VerifierInputError('domain_mismatch', 'domain_separator')
    }
  }

  return {
    revocationRoot: fields[0],
    nullifier: fields[1],
    domainSeparator: domain,
    credentialRoot: fields[3],
  }
}

/** Parse `redaction_witness/v1` public inputs in canonical check order. */
export function parseRedactionWitnessInputs(
  publicInputs: Uint8Array,
): RedactionWitnessInputs {
  const fields = splitFields(publicInputs, REDACTION_PUBLIC_INPUTS_LEN, REDACTION_FIELD_COUNT)

  requireCanonical(fields, REDACTION_FIELDS)

  requireNonZero(fields[0], 'parent_commitment')
  requireNonZero(fields[1], 'output_commitment')
  requireNonZero(fields[2], 'operation_type')
  requireNonZero(fields[3], 'replay_binding')

  const expectedDomain = decodeHex(REDACTION_WITNESS_DOMAIN_TAG_HEX, 'domain_tag')
  const domain = fields[4]
  for (let index = 0; index < FIELD_LEN; index += 1) {
    if (domain[index] !== expectedDomain[index]) {
      throw new VerifierInputError('domain_mismatch', 'domain_tag')
    }
  }

  return {
    parentCommitment: fields[0],
    outputCommitment: fields[1],
    operationType: fields[2],
    replayBinding: fields[3],
    domainTag: domain,
  }
}

/** Dispatch to the parser for `schema`. */
export function parsePublicInputs(
  schema: string,
  publicInputs: Uint8Array,
): SilentWitnessInputs | RevocationWitnessInputs | RedactionWitnessInputs {
  if (schema === SCHEMA_SILENT_WITNESS) {
    return parseSilentWitnessInputs(publicInputs)
  }
  if (schema === SCHEMA_REVOCATION_WITNESS) {
    return parseRevocationWitnessInputs(publicInputs)
  }
  if (schema === SCHEMA_REDACTION_WITNESS) {
    return parseRedactionWitnessInputs(publicInputs)
  }
  throw new VerifierInputError('unknown_schema', 'schema')
}

/**
 * Classify one conformance case. Returns `null` when the material is accepted,
 * otherwise the stable reject code. The check order — public inputs first,
 * then the proof blob — is part of the codec contract.
 */
export function classify(
  schema: string,
  publicInputsHex: string,
  proofHex: string,
): RejectCode | null {
  try {
    const publicInputs = decodeHex(publicInputsHex, 'public_inputs', 'length')
    parsePublicInputs(schema, publicInputs)
    const proof = decodeHex(proofHex, 'proof')
    checkProofBounds(proof)
  } catch (error) {
    if (error instanceof VerifierInputError) {
      return error.code
    }
    throw error
  }
  return null
}
